### PR TITLE
[move-vm] separate error code for vector ops

### PR DESCRIPTION
**Pull request by mengxu-fb**
The `ABORTED` error code is reserved for the Move `abort` operation.
Runtime errors originated from vector operations should have their own
error category (similar to `ARITHMETIC_ERROR`). This commit adds the new
error category `VECTOR_OPERATION_ERROR`.

NOTE: this change applies to the vector bytecode only but not the native
functions in move-stdlib. These native functions still return `ABORTED` with
a sub error code. This is achieved by mapping `VECTOR_OPERATION_ERROR`
into `ABORTED` at the native function adapter layer.

~~NOTE: a few test cases in move-stdlib `VectorTests.move` need to take
the following adjustment:~~

~~```#[expected_failure(abort_code = 1)``` --> ```#[expected_failure]```~~

~~This is due to a limitation documented in #8338 that the unit-test result
checker only checks sub error code when the major error code is `ABORTED`.
Now with `VECTOR_OPERATION_ERROR` being the major error code,
these sub code won't be caught by the unit-test framework, which is similar
to the situation of `ARITHMETIC_ERROR` discussed in the issue. Double
checking with @tzakian to see whether this needs to be address in this PR.~~

## Motivation

Consistency of error codes

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

CI, especially the test cases in the move-vm

### CODE REVIEW

**Review by tnowacki — CHANGES_REQUESTED** [review]
We shouldn't change the behavior of the native functions, just the new bytecode operations. So the VectorTests.move should remain unchanged

**Comment by mengxu-fb** [comment]
Fixed. The native functions' behaviors are reverted back, they still return `ABORTED` now.

File: language/move-stdlib/src/natives/vector.rs [hunk 1]
```diff
@@ -109,5 +116,26 @@ pub fn native_swap(
...
     let cost = native_gas(context.cost_table(), NativeCostIndex::SWAP, 1);
-    NativeResult::map_partial_vm_result_empty(cost, r.swap(idx1, idx2, &ty_args[0]))
+    NativeResult::map_partial_vm_result_empty(
+        cost,
+        r.swap(idx1, idx2, &ty_args[0])
+            .map_err(native_error_to_abort),
+    )
+}
+
+fn native_error_to_abort(err: PartialVMError) -> PartialVMError {
```
Comment (tnowacki) [inline language/move-stdlib/src/natives/vector.rs:126]: This feels a little gross, but we are deprecating the vector natives eventually so probably best to keep the grossness contained to this file

**Review by tnowacki — APPROVED** [review]

File: language/move-stdlib/src/natives/vector.rs (hunk 1 above)
Comment (mengxu-fb) [reply language/move-stdlib/src/natives/vector.rs:126]: I agree. It's a hack and hopefully will go away when we deprecate the vector natives.

**Pull request merged**